server/src/main/java/org/apache/iotdb/db/tools/settle/TsFileAndModSettleTool.java

```diff
@@ -0,0 +1,353 @@
+      TsFileResource resourceToBeSettled, List<TsFileResource> settledResources) throws Exception {
+    if (!resourceToBeSettled.isClosed()) {
+      logger.warn(
+          "The tsFile {} should be sealed when rewritting.", resourceToBeSettled.getTsFilePath());
+      return;
+    }
+    // if no deletions to this tsfile, then return.
+    if (!resourceToBeSettled.getModFile().exists()) {
+      return;
+    }
+    try (TsFileRewriteTool tsFileRewriteTool = new TsFileRewriteTool(resourceToBeSettled)) {
+      tsFileRewriteTool.parseAndRewriteFile(settledResources);
+    }
+    if (settledResources.size() == 0) { // if all the data in this tsfile has been deleted
+      resourceToBeSettled.readUnlock();
+      resourceToBeSettled.writeLock();
+      resourceToBeSettled.delete();
+      resourceToBeSettled.writeUnlock();
+      resourceToBeSettled.readLock();
+    }
+    return;
+  }
+
+  public static void findFilesToBeRecovered() {
+    if (FSFactoryProducer.getFSFactory().getFile(SettleLog.getSettleLogPath()).exists()) {
+      try (BufferedReader settleLogReader =
+          new BufferedReader(
+              new FileReader(
+                  FSFactoryProducer.getFSFactory().getFile(SettleLog.getSettleLogPath())))) {
+        String line = null;
+        while ((line = settleLogReader.readLine()) != null && !line.equals("")) {
+          String oldFilePath = line.split(SettleLog.COMMA_SEPERATOR)[0];
+          Integer settleCheckStatus = Integer.parseInt(line.split(SettleLog.COMMA_SEPERATOR)[1]);
+          if (settleCheckStatus == SettleCheckStatus.SETTLE_SUCCESS.getCheckStatus()) {
+            recoverSettleFileMap.remove(oldFilePath);
+            continue;
+          }
+          recoverSettleFileMap.put(oldFilePath, settleCheckStatus);
+        }
+      } catch (IOException e) {
+        logger.error(
+            "meet error when recover settle process, file path:{}",
+            SettleLog.getSettleLogPath(),
+            e);
+      }
+    }
+  }
+
+  /** this method is used to check whether the new file is settled when recovering old tsFile. */
+  public boolean isSettledFileGenerated(TsFileResource oldTsFileResource) {
+    String oldFilePath = oldTsFileResource.getTsFilePath();
+    return TsFileAndModSettleTool.recoverSettleFileMap.containsKey(oldFilePath)
+        && TsFileAndModSettleTool.recoverSettleFileMap.get(oldFilePath)
+            == SettleCheckStatus.AFTER_SETTLE_FILE.getCheckStatus();
+  }
+
+  /** when the new file is settled , we need to find and deserialize it. */
+  public List<TsFileResource> findSettledFile(TsFileResource resourceToBeSettled)
+      throws IOException {
+    List<TsFileResource> settledTsFileResources = new ArrayList<>();
+    SettleLog.writeSettleLog(
+        resourceToBeSettled.getTsFilePath()
+            + SettleLog.COMMA_SEPERATOR
+            + SettleCheckStatus.BEGIN_SETTLE_FILE);
+
+    for (File tempPartitionDir : resourceToBeSettled.getTsFile().getParentFile().listFiles()) {
```

Comment:
Dereference of `resourceToBeSettled.getTsFile().getParentFile().listFiles()` may produce `NullPointerException`.